pd/src/app.rs

```diff
@@ -335,29 +333,106 @@ impl App {
-            .lock()
-            .unwrap()
-            .set_height(end.height);
-
-        // TODO: if necessary, set the EndBlock response to add validators
-        // at the epoch boundary
-        if end.height.unsigned_abs() == epoch.start_height().value() {
-            // Epoch boundary -- add/remove validators if necessary
-            tracing::info!("new epoch");
-            increment_counter!("epoch");
+            .clone();
+
+        let height: u64 = end.height.try_into().expect("height should be nonnegative");
+        let epoch = pending_block.lock().unwrap().set_height(height);
+
+        let state = self.state.clone();
+        async move {
+            if epoch.end_height().value() == height {
+                tracing::info!(
+                    ?height,
+                    ?epoch,
+                    "last block of epoch, processing rate updates"
+                );
+                metrics::increment_counter!("epoch");
+
+                tracing::debug!("fetching rate data from state");
+                let current_base_rate = state.base_rate_data(epoch.index).await?;
+                let current_rates = state.rate_data(epoch.index).await?;
+                tracing::debug!(?current_base_rate);
+                for current_rate in &current_rates {
+                    tracing::debug!(?current_rate);
+                }
+
+                // here's where we should compute updated rates (moved from state code)
+                // TODO @ava insert calls here
+
+                // move below into calls to methods in stake crate?
+
+                // 1 bps = 1e-4, so here we group digits by 4s rather than 3s as is usual
+                let base_reward_rate = 3_0000; // 3bps -> 11% return over 365 epochs, why not
+                let base_exchange_rate = (current_base_rate.base_exchange_rate
+                    * (base_reward_rate + 1_0000_0000))
+                    / 1_0000_0000;
+                let next_base_rate = BaseRateData {
+                    base_exchange_rate,
+                    base_reward_rate,
+                    epoch_index: epoch.index + 1,
+                };
+
+                let next_rates = current_rates
+                    .iter()
+                    .map(|current_rate| {
+                        // TODO (hdevalence) use funding streams here, this ignores funding streams
+                        let validator_reward_rate = base_reward_rate; // (minus fs term)
+                        let validator_exchange_rate = (current_rate.validator_exchange_rate
+                            * (validator_reward_rate + 1_0000_0000))
+                            / 1_0000_0000;
+
+                        // this is supposed to be multiplied by the number of delegation tokens,
+                        // how do we track that?
+                        let _voting_power_adjustment =
+                            (validator_exchange_rate * 1_0000_0000) / base_exchange_rate;
```

Comment:
We can compute the voting power adjustment factor, but we can't compute the voting power itself without knowing the number of delegation tokens for that validator.  Rather than doing something special for tracking delegation tokens, it seems useful to track the supply of all tokens.

For now, this leaves the voting power unchanged (in the application level, it's also unchanged in the consensus level).